### PR TITLE
Added code to pass appropriate logs to on_test|train_end() methods

### LINKED ISSUE #38498: Currently logs param is None for on_train_end() and on_test_end()
You can see the current implementation of `fit()` ([here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/engine/training.py#L950)) and `evaluate()` ([here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/engine/training.py#L1180)) methods that the logs passed to the methods `on_train_end()` and `on_test_end()` are None which is as per the documentation can be changed in future.  
In tensorflow/addons, There is addition of `TQDMProgressBar()` callback. And recently I have raised PR [#1649](https://github.com/tensorflow/addons/pull/1649) to add code to make progress bar work in case of `evaluate()` too.  
Here, we came across the problem that there are `logs` passed to `on_test_batch_end()` method to update the progress bar. But after the epoch is complete and when `on_test_end()` method is called, there are no `logs` passed to that. Because of this, there is no metrics results passed to the method. But in my opinion and also from @shun-lin's [#1649 (comment)](https://github.com/tensorflow/addons/pull/1649#issuecomment-612785521), it is good to pass logs which are output from the last call to `on_test_batch_end()` method. Currenly in tqdm callback, we are storing the `on_test_batch_end()` logs in class variable and using them in `on_test_end()`, which we think is temporary fix. 

cc @shun-lin, @gabrieldemarmiesse.

### PR DESCRIPTION
**Pull request by ashutosh1919**
fixes #38498 .  
@mihaimaruseac , please review.

### CODE REVIEW

File: tensorflow/python/keras/engine/training.py [hunk 1]
```diff
@@ -944,10 +945,11 @@ def fit(self,
           epoch_logs.update(val_logs)
 
         callbacks.on_epoch_end(epoch, epoch_logs)
+        training_logs = epoch_logs
         if self.stop_training:
           break
 
-      callbacks.on_train_end()
+      callbacks.on_train_end(training_logs)
```
Comment (mihaimaruseac) [inline tensorflow/python/keras/engine/training.py:952]:
```suggestion
      callbacks.on_train_end(logs=training_logs)
```

This way even reordering the arguments will keep the semantics

File: tensorflow/python/keras/engine/training.py [hunk 2]
```diff
@@ -1177,9 +1179,9 @@ def evaluate(self,
               logs = tmp_logs  # No error, now safe to assign to logs.
               end_step = step + data_handler.step_increment
               callbacks.on_test_batch_end(end_step, logs)
-      callbacks.on_test_end()
-
       logs = tf_utils.to_numpy_or_python_type(logs)
+      callbacks.on_test_end(logs)
```
Comment (mihaimaruseac) [inline tensorflow/python/keras/engine/training.py:1183]:
```suggestion
      callbacks.on_test_end(logs=logs)
```

**Review by mihaimaruseac — COMMENTED** [review]
LGTM but I'll get some review from Keras owners too

**Review by mihaimaruseac — APPROVED** [review]

**Review by mihaimaruseac — APPROVED** [review]

**Comment by ashutosh1919** [comment]
@mihaimaruseac , I am not able to see the logs of which part is failed. Can you paste the stack trace here? Thanks!

**Comment by mihaimaruseac** [comment]
Please ignore the macospy2 test. The windows failure is at https://source.cloud.google.com/results/invocations/c496c74d-d468-4265-bab5-f2a95242dcd0/targets/%2F%2Ftensorflow%2Ftools%2Fci_build%2Fbuilds:gen_win_out/log

**Comment by ashutosh1919** [comment]
@mihaimaruseac , I don't think the error in windows Bazel CPU is because of the code added in this PR. Please check once and give your comments.

**Comment by ashutosh1919** [comment]
@mihaimaruseac , when I am trying to rebase this PR with tf master branch, it is giving conflict errors in files which are not even changed. Please consider this fact too.

**Comment by mihaimaruseac** [comment]
Triggered a new build.

Regarding the rebase, how do you do it? Do you have 2 upstreams (your fork and tensorflow)? That would be the recommended way as it works in most cases:

```console
$ git clone ${your_fork_url}
$ cd tensorflow
$ git remote add upstream ${main_tensorflow_repo_url}
$ # do work, git add, git commit, everything on a branch for your repo
$ git push # -u optionally
$ # now syncing from upstream
$ git checkout master
$ git pull --rebase upstream
$ git push  # now master on your fork is synced with master upstream
$ git checkout -  # switch back to the branch
$ git rebase master  # rebase on master without any merge commits
$ git push  # now the changes on your repo and rebased on master upstream
```

**Comment by ashutosh1919** [comment]
@mihaimaruseac , Thanks for the code snippet. I actually had only one upstream. You cleared my doubts on that. Thanks so much!

**Pull request merged**